src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedArray`1.cs

```diff
@@ -0,0 +1,369 @@
+            }
+
+            return false;
+        }
+
+        bool ICollection<T>.Contains(T value)
+        {
+            foreach (ICollection<T> collection in _items)
+            {
+                if (collection.Contains(value))
+                    return true;
+            }
+
+            return false;
+        }
+
+        int IList.IndexOf(object value)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                IList list = _items[i];
+                var index = list.IndexOf(value);
+                if (index >= 0)
+                {
+                    return index + i * s_segmentSize;
+                }
+            }
+
+            return -1;
+        }
+
+        int IList<T>.IndexOf(T value)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                IList<T> list = _items[i];
+                var index = list.IndexOf(value);
+                if (index >= 0)
+                {
+                    return index + i * s_segmentSize;
+                }
+            }
+
+            return -1;
+        }
+
+        void IList.Insert(int index, object value)
+        {
+            IList list = Array.Empty<T>();
+            list.Insert(0, value);
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        void IList<T>.Insert(int index, T value)
+        {
+            IList<T> list = Array.Empty<T>();
+            list.Insert(0, value);
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        void IList.Remove(object value)
+        {
+            IList list = Array.Empty<T>();
+            list.Remove(value);
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        bool ICollection<T>.Remove(T value)
+        {
+            ICollection<T> collection = Array.Empty<T>();
+            _ = collection.Remove(value);
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        void IList.RemoveAt(int index)
+        {
+            IList list = Array.Empty<T>();
+            list.RemoveAt(0);
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        void IList<T>.RemoveAt(int index)
+        {
+            IList<T> list = Array.Empty<T>();
+            list.RemoveAt(0);
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+            => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator()
+            => GetEnumerator();
+
+        int IStructuralComparable.CompareTo(object other, IComparer comparer)
+        {
+            if (other is null)
+                return 1;
+
+            if (!(other is SegmentedArray<T> o))
+            {
+                // Delegate to T[] so the correct exception is thrown
+                IStructuralComparable comparable = Array.Empty<T>();
+                comparable.CompareTo(new object(), comparer);
+                throw ExceptionUtilities.Unreachable;
```

Comment:
i don't really understand what's going on here, or why this would be correct.

Reply:
Fixed